File: frontend/src/types/swarm.ts
export interface Agent {
  name: string;
  system_prompt: string;
  tools: string[];
  model_id?: string;
  description?: string;
  icon?: string;
}

export interface SwarmExecutionRequest {
  task: string;
  agents: Agent[];
  max_handoffs?: number;
  max_iterations?: number;
  execution_timeout?: number;
  node_timeout?: number;
  execution_id?: string;
  background?: boolean;
}

export interface SwarmExecutionResult {
  execution_id: string;
  status: 'completed' | 'failed' | 'stopped';
  result?: string;
  error?: string;
  execution_time?: number;
  handoffs: number;
  tokens_used: number;
  agent_sequence: string[];
  artifacts: Artifact[];
}

export interface SwarmEvent {
  type:
      | "execution_started"
      | "execution_completed"
      | "execution_failed"
      | "execution_stopped"
      | "agent_started"
      | "agent_completed"
      | "text_generation"
      | "tool_use"           // When agent decides to use a tool
      | "tool_execution"     // When tool is actually being executed
      | "tool_called"        // Alternative name for tool execution
      | "tool_approval_required"  // NEW: Tool needs approval (execution paused)
      | "tool_approval_response"   // NEW: Approval granted/denied (execution resumed)
      | "tool_rejected"      // NEW: Tool was rejected
      | "tool_executed"      // NEW: Tool successfully executed after approval
      | "handoff"
      | "handoff_requested"
      | "handoff_approved"
      | "system_message"
      | "artifact_created"
      | "approval_required"
      | "safety_flag"
      | "metrics_updated"
      | "execution_paused"
      | "execution_resumed"
      | "plan_created"
      | "step_started"
      | "step_completed"
      | "action_started"
      | "action_completed"
      | "error";
  timestamp: string | Date;
  agent?: string;
  data?: any;
  executionId?: string;
  stepId?: string;
  actionId?: string;
}

export interface Artifact {
  id: string;
  type: 'code' | 'html' | 'document' | 'data' | 'image' | 'test_result';
  title: string;
  content: string;
  language?: string;
  filename?: string;
  agentId?: string;
  timestamp?: Date;
  size?: number;
  metadata?: {
    language?: string;
    agent?: string;
    lines?: number;
    functions?: number;
    [key: string]: any;
  };
}

export interface ExecutionMetrics {
  totalAgents: number;     // Changed from totalDuration
  handoffs: number;        // Changed from totalHandoffs
  toolUses: number;        // Changed from totalToolCalls
  events: number;          // New field
  tokensUsed: number;      // Changed from totalTokens
  executionTime: number;   // Changed from totalDuration
}

export interface AgentTemplate {
  id: string;
  name: string;
  description: string;
  icon: string;
  system_prompt: string;
  tools: string[];
  category: 'research' | 'development' | 'analysis' | 'creative' | 'review';
}

File: backend/app/schemas/swarm.py
# schemas/swarm.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum


class ExecutionStatus(str, Enum):
    QUEUED = "queued"
    STARTED = "started"
    COMPLETED = "completed"
    FAILED = "failed"
    STOPPED = "stopped"


class AgentConfig(BaseModel):
    name: str
    system_prompt: str
    tools: List[str] = []
    description: Optional[str] = None
    icon: Optional[str] = None
    model: Optional[str] = "gpt-4o-mini"  # Default model
    temperature: Optional[float] = 0.7
    max_tokens: Optional[int] = 16000
    
    class Config:
        # Fix the model_id field warning
        protected_namespaces = ()


class AddAgentRequest(BaseModel):
    session_id: Optional[str] = None
    agent_name: str = Field(..., min_length=1, max_length=100)
    agent_role: str = Field(..., min_length=1, max_length=200)
    agent_description: Optional[str] = Field(None, max_length=500)
    tools: List[str] = []
    model: Optional[str] = "gpt-4o-mini"
    temperature: Optional[float] = 0.7


class AddAgentResponse(BaseModel):
    success: bool
    agent_id: str
    message: str


class SwarmExecutionRequest(BaseModel):
    task: str
    agents: Optional[List[AgentConfig]] = None
    max_handoffs: Optional[int] = 10
    max_iterations: Optional[int] = 50
    execution_timeout: Optional[int] = 300
    node_timeout: Optional[int] = 60
    execution_id: Optional[str] = None
    session_id: Optional[str] = None  # For human-loop compatibility
    background: bool = False
    use_orchestrator: Optional[bool] = True
    context: Optional[Dict[str, Any]] = None  # For preserving session context
    execution_mode: Optional[str] = "auto"  # "sequential", "parallel", or "auto"


class Artifact(BaseModel):
    type: str
    name: str
    content: str
    metadata: Optional[Dict[str, Any]] = None


class SwarmExecutionResponse(BaseModel):
    execution_id: str
    status: ExecutionStatus
    result: Optional[str] = None
    error: Optional[str] = None
    execution_time: Optional[float] = None
    handoffs: int = 0
    tokens_used: int = 0
    agent_sequence: List[str] = []
    artifacts: List[Artifact] = []
    message: Optional[str] = None
    execution_mode: Optional[str] = None  # "sequential" or "parallel"
    execution_details: Optional[Dict[str, Any]] = None  # Mode-specific details


class SwarmEvent(BaseModel):
    type: str
    timestamp: str
    agent: Optional[str] = None
    data: Optional[Dict[str, Any]] = None


class SwarmListResponse(BaseModel):
    executions: List[SwarmExecutionResponse]


class AgentTemplateResponse(BaseModel):
    name: str
    description: str
    tools: List[str]
    system_prompt: str